sapi/fuzzing/fuzzer-parser.c

```diff
@@ -0,0 +1,101 @@
+/*
+   +----------------------------------------------------------------------+
+   | PHP Version 7                                                        |
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Johannes Schlüter <johanes@php.net>                         |
+   |          Stanislav Malyshev <stas@php.net>                           |
+   +----------------------------------------------------------------------+
+ */
+
+#include <main/php.h>
+#include <main/php_main.h>
+#include <main/SAPI.h>
+#include <ext/standard/info.h>
+#include <ext/standard/php_var.h>
+#include <main/php_variables.h>
+#ifdef JO0
+#include <ext/standard/php_smart_str.h>
+#endif
+
+#include "fuzzer.h"
+
+#include "fuzzer-sapi.h"
+#include "compat.h"
+
+int fuzzer_do_parse(zend_file_handle *file_handle, char *filename)
+{
+	int retval = FAILURE; /* failure by default */
+
+	SG(options) |= SAPI_OPTION_NO_CHDIR;
+	SG(request_info).argc=0;
+	SG(request_info).argv=NULL;
+
+	if (php_request_startup(TSRMLS_C)==FAILURE) {
+		php_module_shutdown(TSRMLS_C);
+		return FAILURE;
+	}
+
+	SG(headers_sent) = 1;
+	SG(request_info).no_headers = 1;
+	php_register_variable("PHP_SELF", filename, NULL TSRMLS_CC);
+
+	zend_first_try {
+		zend_compile_file(file_handle, ZEND_REQUIRE);
+		//retval = php_execute_script(file_handle TSRMLS_CC);
+	} zend_end_try();
+
+	php_request_shutdown((void *) 0);
+
+	return (retval == SUCCESS) ? SUCCESS : FAILURE;
+}
+
+int fuzzer_do_request(zend_file_handle *file_handle, char *filename);
+int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
+	char *s = malloc(Size+1);
+	memcpy(s, Data, Size);
+	s[Size] = '\0';
```

Comment:
Don't we need MMAP_AHEAD here? I'm pretty surprised that this works.

Reply:
Not sure why we'd need MMAP_AHEAD - this looks like pretty standard code. Am I missing anything?

Reply:
The lexer needs a certain number of bytes after the end of the input to be allocated to perform lookahead. This is generally ensured automatically by the stream logic -- as you are explicitly setting the buffer here though, it will be bypassed. I would expect that this results in asan out-of-bounds read reports.

Reply:
That sounds a bit weird - shouldn't lexer stop at the end of the data, at \0? Reading past that would be reading unallocated memory, regardless of the source. Am I still missing something here? Regardless of this patch, reading past allocated buffer is not valid, lookahead or not.